File: data/derived_cot/rq3_prompting/results/translation/r1/java_to_py/code_output/UserLoginDB.py
import sqlite3


class UserLoginDB:
    def __init__(self, db_name: str):
        self.connection = None
        try:
            self.connection = sqlite3.connect(db_name)
            self._create_table()
        except sqlite3.Error as e:
            print(e)

    def _create_table(self):
        create_table_query = (
            "CREATE TABLE IF NOT EXISTS users (username TEXT, password TEXT)"
        )
        try:
            cursor = self.connection.cursor()
            cursor.execute(create_table_query)
            self.connection.commit()
        except sqlite3.Error as e:
            print(e)

    def insert_user(self, username: str, password: str):
        insert_query = "INSERT INTO users (username, password) VALUES (?, ?)"
        try:
            cursor = self.connection.cursor()
            cursor.execute(insert_query, (username, password))
            self.connection.commit()
        except sqlite3.Error as e:
            print(e)

    def search_user_by_username(self, username: str):
        search_query = "SELECT * FROM users WHERE username = ?"
        try:
            cursor = self.connection.cursor()
            cursor.execute(search_query, (username,))
            row = cursor.fetchone()
            if row:
                return f"{row[0]},{row[1]}"
        except sqlite3.Error as e:
            print(e)
        return None

    def delete_user_by_username(self, username: str):
        delete_query = "DELETE FROM users WHERE username = ?"
        try:
            cursor = self.connection.cursor()
            cursor.execute(delete_query, (username,))
            self.connection.commit()
        except sqlite3.Error as e:
            print(e)

    def validate_user_login(self, username: str, password: str) -> bool:
        user = self.search_user_by_username(username)
        if user:
            parts = user.split(",")
            return parts[1] == password
        return False

    def close(self):
        try:
            if self.connection:
                self.connection.close()
        except sqlite3.Error as e:
            print(e)

File: data/derived_cot/rq1_traces/translation/glm5.1/java_to_cpp/code_output/UserLoginDB.cpp
#ifndef USER_LOGIN_DB_H
#define USER_LOGIN_DB_H

#include <sqlite3.h>
#include <string>
#include <optional>
#include <iostream>
#include <sstream>
#include <vector>

class UserLoginDB {
private:
    sqlite3* connection;

    void createTable() {
        const char* createTableQuery = "CREATE TABLE IF NOT EXISTS users (username TEXT, password TEXT)";
        char* errMsg = nullptr;
        int rc = sqlite3_exec(connection, createTableQuery, nullptr, nullptr, &errMsg);
        if (rc != SQLITE_OK) {
            std::cerr << errMsg << std::endl;
            sqlite3_free(errMsg);
        }
    }

public:
    UserLoginDB(const std::string& dbName) : connection(nullptr) {
        int rc = sqlite3_open(dbName.c_str(), &connection);
        if (rc == SQLITE_OK) {
            createTable();
        } else {
            std::cerr << sqlite3_errmsg(connection) << std::endl;
            sqlite3_close(connection);
            connection = nullptr;
        }
    }

    void insertUser(const std::string& username, const std::string& password) {
        const char* insertQuery = "INSERT INTO users (username, password) VALUES (?, ?)";
        sqlite3_stmt* pstmt = nullptr;
        int rc = sqlite3_prepare_v2(connection, insertQuery, -1, &pstmt, nullptr);
        if (rc != SQLITE_OK) {
            std::cerr << sqlite3_errmsg(connection) << std::endl;
            return;
        }
        sqlite3_bind_text(pstmt, 1, username.c_str(), -1, SQLITE_TRANSIENT);
        sqlite3_bind_text(pstmt, 2, password.c_str(), -1, SQLITE_TRANSIENT);
        rc = sqlite3_step(pstmt);
        if (rc != SQLITE_DONE) {
            std::cerr << sqlite3_errmsg(connection) << std::endl;
        }
        sqlite3_finalize(pstmt);
    }

    std::optional<std::string> searchUserByUsername(const std::string& username) {
        const char* searchQuery = "SELECT * FROM users WHERE username = ?";
        sqlite3_stmt* pstmt = nullptr;
        int rc = sqlite3_prepare_v2(connection, searchQuery, -1, &pstmt, nullptr);
        if (rc != SQLITE_OK) {
            std::cerr << sqlite3_errmsg(connection) << std::endl;
            return std::nullopt;
        }
        sqlite3_bind_text(pstmt, 1, username.c_str(), -1, SQLITE_TRANSIENT);
        rc = sqlite3_step(pstmt);
        if (rc == SQLITE_ROW) {
            const char* usernameCol = reinterpret_cast<const char*>(sqlite3_column_text(pstmt, 0));
            const char* passwordCol = reinterpret_cast<const char*>(sqlite3_column_text(pstmt, 1));
            // Java's getString returns null for SQL NULL; concatenation converts null to "null"
            std::string usernameStr = usernameCol ? std::string(usernameCol) : "null";
            std::string passwordStr = passwordCol ? std::string(passwordCol) : "null";
            std::string result = usernameStr + "," + passwordStr;
            sqlite3_finalize(pstmt);
            return result;
        }
        sqlite3_finalize(pstmt);
        return std::nullopt;
    }

    void deleteUserByUsername(const std::string& username) {
        const char* deleteQuery = "DELETE FROM users WHERE username = ?";
        sqlite3_stmt* pstmt = nullptr;
        int rc = sqlite3_prepare_v2(connection, deleteQuery, -1, &pstmt, nullptr);
        if (rc != SQLITE_OK) {
            std::cerr << sqlite3_errmsg(connection) << std::endl;
            return;
        }
        sqlite3_bind_text(pstmt, 1, username.c_str(), -1, SQLITE_TRANSIENT);
        rc = sqlite3_step(pstmt);
        if (rc != SQLITE_DONE) {
            std::cerr << sqlite3_errmsg(connection) << std::endl;
        }
        sqlite3_finalize(pstmt);
    }

    bool validateUserLogin(const std::string& username, const std::string& password) {
        auto user = searchUserByUsername(username);
        if (user.has_value()) {
            // Replicate Java's String.split(",") — split on every comma, take index 1
            std::vector<std::string> parts;
            std::istringstream iss(*user);
            std::string part;
            while (std::getline(iss, part, ',')) {
                parts.push_back(part);
            }
            if (parts.size() > 1) {
                return parts[1] == password;
            }
        }
        return false;
    }

    void close() {
        if (connection != nullptr) {
            sqlite3_close(connection);
            connection = nullptr;
        }
    }

    ~UserLoginDB() {
        close();
    }

    // Prevent copying to avoid double-close
    UserLoginDB(const UserLoginDB&) = delete;
    UserLoginDB& operator=(const UserLoginDB&) = delete;
};

#endif // USER_LOGIN_DB_H